be/src/exprs/agg/map_agg.h

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <fmt/format.h>
+
+#include "column/binary_column.h"
+#include "column/column.h"
+#include "column/column_helper.h"
+#include "column/fixed_length_column.h"
+#include "column/hash_set.h"
+#include "column/map_column.h"
+#include "column/type_traits.h"
+#include "exprs/agg/aggregate.h"
+#include "exprs/function_context.h"
+#include "gutil/casts.h"
+#include "util/phmap/phmap.h"
+#include "util/time.h"
+
+namespace starrocks {
+
+template <LogicalType KT, typename MyHashMap = std::map<int, Datum>>
+struct MapAggAggregateFunctionState : public AggregateFunctionEmptyState {
+    using KeyColumnType = RunTimeColumnType<KT>;
+    using KeyType = typename SliceHashSet::key_type;
+
+    MyHashMap hash_map;
+    void update(MemPool* mem_pool,
+                const KeyColumnType& arg_key_column,
+                const Column& arg_value_column,
+                size_t offset, size_t count) {
+        for (int i = offset; i < offset + count; i++) {
+            if constexpr (!lt_is_string<KT>) {
+                auto key = arg_key_column.get_data()[i];
+                auto value = arg_value_column.get(i);
+                hash_map.emplace(key, value);
+            } else {
+                auto raw_key = arg_key_column.get_slice(i);
+                KeyType key(raw_key);
+                if (!hash_map.contains(key)) {
+                    uint8_t* pos = mem_pool->allocate(key.size);
+                    memcpy(pos, key.data, key.size);
+                    hash_map.emplace(Slice(pos, key.size), arg_value_column.get(i));
+                }
+            }
+        }
+    }
+};
+
+template <LogicalType KT, typename MyHashMap = std::map<int, int>>
+class MapAggAggregateFunction final
+        : public AggregateFunctionBatchHelper<MapAggAggregateFunctionState<KT, MyHashMap>,
+                                              MapAggAggregateFunction<KT, MyHashMap>> {
+public:
+    using KeyColumnType = RunTimeColumnType<KT>;
+    using CppType = RunTimeCppType<KT>;
+
+    void update(FunctionContext* ctx, const Column** columns, AggDataPtr __restrict state,
+                size_t row_num) const override {
+        // Key could not be null.
+        if ((columns[0]->is_nullable() && columns[0]->is_null(row_num)) || columns[0]->only_null()) {
+            return;
+        }
```

Comment:
these code could be removed

Reply:
This is to skip the NULL value of key column. 
I removed this and execute the sql above, the result will be:
![image](https://github.com/user-attachments/assets/c5bae64b-3a8b-457b-9bc7-1d87511b8956)

Reply:
you can regist it by add_aggregate_mapping. The framework will help you ignore the null elements

Reply:
I used to regist it by add_aggregate_mapping, but inside this function, it is actually used AggregateFactory::MakeNullableAggregateFunctionUnary, which only transfer the first column to nest function and   map_agg need two arguments.

Reply:
MakeNullableAggregateFunctionVariadic could accept more than 1 column, but it skip the rows with null value. In map_agg, key can not be null but value can be null. So both the frameworks can not be used.